src/core/thread/mlr_types.hpp

```diff
@@ -0,0 +1,64 @@
+/*
+ *  Copyright (c) 2020, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file includes definitions for Multicast Listener Registration states of multicast addresses.
+ */
+
+#ifndef MLR_TYPES_HPP_
+#define MLR_TYPES_HPP_
+
+#include "openthread-core-config.h"
+
+#if (OPENTHREAD_FTD || OPENTHREAD_MTD) && OPENTHREAD_CONFIG_MLR_ENABLE
+
+namespace ot {
+
+/**
+ * @addtogroup core-mlr
+ *
+ * @{
+ *
+ */
+
+/**
+ * Multicast Listener Registration state for multicast addresses.
+ *
+ */
+typedef enum MlrState
```

Comment:
nit: prefer to define explicit type

Reply:
Defining it to explicit type `uint8_t` could lead to compile issues:

```
2020-07-21T02:24:12.9702646Z /tmp/ot-size-report/b/examples/../src/core/net/netif.hpp:177:26: error: 'ot::Ip6::NetifMulticastAddress::mMlrState' is too small to hold all values of 'ot::MlrState {aka enum ot::MlrState}' [-Werror]
2020-07-21T02:24:12.9702794Z      MlrState mMlrState : 2;
```

Same issue happened to one of @librasungirl 's PR too.